pipeline_dp/private_beam.py

```diff
@@ -308,3 +315,146 @@ def __init__(self, fn: Callable, label: Optional[str] = None):
+        stored in a field of 'self'. It will be serialized and distributed
+        to the workers together with 'self'.
+
+        Warning: do not store 'budget_accountant' in 'self'. It is assumed to
+        live in the driver process.
+        """
+        pass
+
+    def set_aggregate_params(self,
+                             aggregate_params: pipeline_dp.AggregateParams):
+        """Sets aggregate parameters
+
+        The custom combiner can optionally use it for own DP parameter
+        computations.
+        """
+        self._aggregate_params = aggregate_params
+
+
+class _CombineFnCombiner(pipeline_dp.CustomCombiner):
+
+    def __init__(self, private_combine_fn: PrivateCombineFn):
+        self._private_combine_fn = private_combine_fn
+
+    def create_accumulator(self, values):
+        """Creates accumulator from 'values'."""
+        accumulator = self._private_combine_fn.create_accumulator()
+        for v in values:
+            accumulator = self._private_combine_fn.add_private_input(
+                accumulator, v)
+        return accumulator
+
+    def merge_accumulators(self, accumulator1, accumulator2):
+        """Merges the accumulators and returns accumulator."""
+        return self._private_combine_fn.merge_accumulators(
+            [accumulator1, accumulator2])
+
+    def compute_metrics(self, accumulator):
+        """Computes and returns the result of aggregation."""
+        return self._private_combine_fn.extract_private_output(accumulator)
+
+    def request_budget(self,
+                       budget_accountant: budget_accounting.BudgetAccountant):
+        self._private_combine_fn.request_budget(budget_accountant)
+
+    def set_aggregate_params(self, aggregate_params):
+        self._private_combine_fn.set_aggregate_params(aggregate_params)
+
+
+@dataclasses.dataclass
+class CombinePerKeyParams:
+    max_partitions_contributed: int
+    max_contributions_per_partition: int
+    budget_weight: float = 1
+    public_partitions: typing.Any = None
+
+
+class CombinePerKey(PrivatePTransform):
+    """Transform class for performing a FlatMap on a PrivatePCollection."""
+
+    def __init__(self,
+                 combine_fn: PrivateCombineFn,
+                 params: CombinePerKeyParams,
+                 label: Optional[str] = None):
+        super().__init__(return_anonymized=True, label=label)
+        self._combine_fn = combine_fn
+        self._params = params
+
+    def expand(self, pcol: pvalue.PCollection):
+        combiner = _CombineFnCombiner(self._combine_fn)
+        aggregate_params = pipeline_dp.AggregateParams(
+            metrics=None,
+            max_partitions_contributed=self._params.max_partitions_contributed,
+            max_contributions_per_partition=self._params.
+            max_contributions_per_partition,
+            custom_combiners=[combiner])
+
+        backend, dp_engine = self._create_dp_engine()
+        # Assumed elements format: (privacy_id, (partition_key, value))
```

Comment:
Should we document that input/output assumptions in the top-level doc for CombinePerKey?